keps/sig-architecture/5241-beta-featuregate-promotion-requirements/README.md

```diff
@@ -0,0 +1,246 @@
+   and that feature can be disabled.
+   Exceptions exist for entirely new APIs and some node features, but this broadly the case.
+3. Alpha means that a feature gate is disabled in all production Kubernetes clusters by default and
+   can be optionally enabled by setting a `--feature-gate` command line argument.
+
+Making the jump to GA (cannot be disabled), without actual field experience is irresponsible.
+The first time we take a feature gate enabled by default in production Kubernetes clusters, we must
+have a way to disable the feature in case of unexpected stability, performance, or security issues.
+
+Enabling incomplete features in production Kubernetes clusters by default is irresponsible.
+Features that are known to be incomplete naturally bring with them additional stability, performance, and security issues.
+Once a feature has been enabled in a production Kubernetes cluster by default, adding to it carries
+greater risk to upgrading clusters and the ecosystem.
+The feature can easily have become relied upon by workloads and other platform extensions.
+If an accident happens in adding those capabilities with stability, performance, and security the
+cost to disable those features in a cluster becomes significantly greater and breaks existing
+clusters, workloads and use-cases.
+This posture makes upgrades higher risk than necessary.
+
+To balance these concerns, we are changing how we evaluate Beta and GA stability criteria.
+The only valid GA criteria are “all issues and gaps identified as feedback during beta are resolved”.
+Promotion from Beta to GA must be zero-diff for the release.
+This means that Beta criteria must include all functional, security, and testing requirements along
+with resolving all issues and gaps identified prior to beta.
+
+Phasing in larger features over time can be done by bringing separate feature gates through alpha, beta, and GA.
+Each feature gate needs to meet the beta and GA criteria for completeness, functional, security, and testing.
+After meeting hte criteria for enabled by default, and at the sig's discretion, the new feature gate could be 
+set to enabled by default in the release it is introduced.
+Importantly, the features need to behave in a way that allows old and new clients to interoperate and new additions
+to larger features able to be independently disablable with their own path for GA.
+
+### Risks and Mitigations
+
+#### What if I need to add capability to my feature?
+To handle this situation, we described above how to add second feature gate for the new behavior.
+This provides a mechanism for adding needed capability, but ensures that
+cluster-admins never end up stuck after upgrade because they rely on v1.Y-1 behavior that new capability
+in v1.Y broke under the same feature gate.
+
+#### Who will make sure that new KEPs follow the promotion rules?
+We'll adjust the KEP template to indicate the allowed criteria, so authors should notice.
```

Comment:
Only for new features. Authors of some existing KEP only notice if they actively sync with the KEP template (rarely done) or we announce this KEP here widely.

Reply:
There's another topic about that in SIG Arch today. [bit.ly/kep-versioning_phase1](https://bit.ly/kep-versioning_phase1)